stablehlo/dialect/ChloBytecode.cpp

```diff
@@ -0,0 +1,284 @@
+/// codes must not be changed, as any changes will break compatibility
+/// with older bytecode.
+///
+/// To add a type, search for "TO ADD TYPE" in this file and ensure each
+/// location is updated.
+enum TypeCode {
+  // TO ADD TYPE: Add an enum value with doc string for new attr.
+
+};
+
+} // namespace chlo_encoding
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// ChloBytecodeInterface
+//===----------------------------------------------------------------------===//
+
+namespace mlir {
+namespace chlo {
+
+namespace {
+/// This class implements the bytecode interface for the chlo dialect.
+class ChloBytecodeInterface : public BytecodeDialectInterface {
+public:
+  ChloBytecodeInterface(Dialect *dialect) : BytecodeDialectInterface(dialect) {}
+
+  //===--------------------------------------------------------------------===//
+  // Attributes
+
+  // These methods are invoked by superclass when an attr from chlo dialect
+  // is encountered.
+  Attribute readAttribute(DialectBytecodeReader &reader) const override;
+  LogicalResult writeAttribute(Attribute attr,
+                               DialectBytecodeWriter &writer) const override;
+
+  // TO ADD ATTRIBUTE: Include a read method for each attribute in CHLO
+  // Ex: SomeAttr readSomeAttr(DialectBytecodeReader &reader) const;
+  ComparisonDirectionAttr
+  readComparisonDirectionAttr(DialectBytecodeReader &reader) const;
+  ComparisonTypeAttr
+  readComparisonTypeAttr(DialectBytecodeReader &reader) const;
+
+  // TO ADD ATTRIBUTE: Include a write method for each attribute in CHLO
+  // Ex: void write(SomeAttr attr, DialectBytecodeWriter &writer) const;
+  void write(ComparisonDirectionAttr attr, DialectBytecodeWriter &writer) const;
+  void write(ComparisonTypeAttr attr, DialectBytecodeWriter &writer) const;
+
+  //===--------------------------------------------------------------------===//
+  // Types
+
+  // These methods are invoked by superclass when a type from chlo dialect
+  // is encountered.
+  Type readType(DialectBytecodeReader &reader) const override;
+  LogicalResult writeType(Type type,
+                          DialectBytecodeWriter &writer) const override;
+
+  // TO ADD TYPE: Include a read method for each type in CHLO
+  // Ex: SomeType readSomeType(DialectBytecodeReader &reader) const;
+
+  // TO ADD TYPE: Include a write method for each type in CHLO
+  // Ex: void write(SomeType attr, DialectBytecodeWriter &writer) const;
+
+private:
+  //===--------------------------------------------------------------------===//
+  // Helper methods
+
+  // Enum reader and writer. Many attrs have a single enum type to serialize.
+  // Use the attributes underlying type to get the numeric value.
+  // Note this may cause issues if enums use an int64_t and have a large value.
+  // All enums in CHLO currently use int32_t.
+  template <typename EnumType, typename EnumTypeAttr, typename SymbolizeFn>
```

Comment:
Actually, how about moving `readEnumAttribute` and `writeEnumAttribute` into Base.h and sharing them between StablehloBytecode and ChloBytecode?

Reply:
I can do that. At some point I also want to relocate some pretty print machinery, likely there if not a separated header. CHLO can leverage the `SameOperandsAndResultType` printing as well